HB32 Vanilla Scheduler Extending the LLVM Generic Scheduler

vb000 opened the pull request:
Custom Scheduler for HB32 Subtarget. Can be enabled when with `hb32sched` option: 

`llc -mcpu=hb-rv32 -hb32sched <...>`

vb000 commented:
For the following unrolled loop:
```c
 #define __remote __attribute__((address_space(1)))

 int vec_add(__remote int* a, __remote int* b, int n) {
   #pragma unroll 8
   for(int i=0; i<n; ++i) {
     a[i] += b[i];
   }

   return 0;
 }
```

I see following results:
```
$ llc -mcpu=hb-rv32 tmp.ll         $ llc -mcpu=hb-rv32 -hb32sched tmp.ll

.LBB0_4:                           .LBB0_4:

  lw  a7, -16(a5)                   lw  a7, -16(a5)
  lw  a3, -16(a2)                   lw  t0, -16(a2)
  add a3, a3, a7                    lw  t1, 12(a2)
  sw  a3, -16(a2)                   lw  t2, 8(a2)
  lw  a7, -12(a5)                   lw  t3, -12(a2)
  lw  a3, -12(a2)                   lw  t4, 4(a2)
  add a3, a3, a7                    lw  t5, 0(a2)
  sw  a3, -12(a2)                   lw  t6, -4(a2)
  lw  a7, -8(a5)                    lw  s1, -8(a2)
  lw  a3, -8(a2)                    add s0, t0, a7
  add a3, a3, a7                    sw  s0, -16(a2)
  sw  a3, -8(a2)                    lw  s0, -12(a5)
  lw  a7, -4(a5)                    addi  a4, a4, 8
  lw  a3, -4(a2)                    add s0, t3, s0
  add a3, a3, a7                    sw  s0, -12(a2)
  sw  a3, -4(a2)                    lw  s0, -8(a5)
  lw  a7, 0(a5)                     add s1, s1, s0
  lw  a3, 0(a2)                     sw  s1, -8(a2)
  add a3, a3, a7                    lw  s1, -4(a5)
  sw  a3, 0(a2)                     add s1, t6, s1
  lw  a7, 4(a5)                     sw  s1, -4(a2)
  lw  a3, 4(a2)                     lw  s1, 0(a5)
  add a3, a3, a7                    add s1, t5, s1
  sw  a3, 4(a2)                     sw  s1, 0(a2)
  lw  a7, 8(a5)                     lw  s1, 4(a5)
  lw  a3, 8(a2)                     add s1, t4, s1
  add a3, a3, a7                    sw  s1, 4(a2)
  sw  a3, 8(a2)                     lw  s1, 8(a5)
  lw  a7, 12(a5)                    add s1, t2, s1
  lw  a3, 12(a2)                    sw  s1, 8(a2)
  add a3, a3, a7                    lw  s1, 12(a5)
  sw  a3, 12(a2)                    addi  a5, a5, 32
  addi  a4, a4, 8                   add s1, t1, s1
  addi  a5, a5, 32                  sw  s1, 12(a2)
  addi  a2, a2, 32                  addi  a2, a2, 32
  bne a6, a4, .LBB0_4               bne a6, a4, .LBB0_4
```

@proftaylor Above is the result I see with unrolled loop. To clarify on the scheduling algorithm we override:
1. Default LLVM scheduler provided two main variants: top-down and bottom-up. We customize default top-down scheduling as opposed to default bottom up because we found [bottom-up is bad for non-blocking loads](https://github.com/vb000/llvm-escapade/blob/master/README-notes.md#test-program). Have you observed the same in GCC?
2. It is run before register allocation on IR in SSA form, which means we have a lot of flexibility in reordering instructions. 
3. This also means the compiler would still reschedule after register allocation based on other heuristics.
4. We only override `load` scheduling --- if there is no load in the top frontier of the control flow graph, we fallback to default top-down scheduling.

Is the counter example you provided still valid?

proftaylor commented:
Thanks Bandhav for this writeup!

A few tips:

a. Before trying to hack the backend,
    i. You should make a hypothesis before you starting mucking with things
about what the optimal code output should be. Scheduling this loop is not a
particularly hard problem, so you should be setting the expectation that it
do as good as hand-optimized code. The code gen you have currently is not
that close to optimal.

    ii. I would suggest trying understand what the scheduler's input is,
and what the algorithm is.  You want to print out the scheduling and/or
aliasing information so you can see why it made the decisions it made, and
by comparing against your hypothesis, you can see what is wrong in its
logic.

b. I would definitely use the restrict keyword on both arrays.  Right now,
I think it is being conservative about moving loads and stores across each
other because it is not sure if the arrays overlap.

c. You may want investigate whether a register renaming pass is being run,
since that was the issue we ran into with GCC.

d. Before you hack the scheduler to prioritize loads, I would just change
the default latency of loads to 20 and see what it does. This give you an
idea of what the current scheduler is capable of.

Cheers!

M

On Thu, Jun 4, 2020 at 8:38 PM Bandhav Veluri <notifications@github.com>
wrote:

> For the following unrolled loop:
>
>  #define __remote __attribute__((address_space(1)))
>
>  int vec_add(__remote int* a, __remote int* b, int n) {
>    #pragma unroll 8
>    for(int i=0; i<n; ++i) {
>      a[i] += b[i];
>    }
>
>    return 0;
>  }
>
> I see following results:
>
> $ llc -mcpu=hb-rv32 tmp.ll         $ llc -mcpu=hb-rv32 -hb32sched tmp.ll
>
> .LBB0_4:                           .LBB0_4:
>
>   lw  a7, -16(a5)                   lw  a7, -16(a5)
>   lw  a3, -16(a2)                   lw  t0, -16(a2)
>   add a3, a3, a7                    lw  t1, 12(a2)
>   sw  a3, -16(a2)                   lw  t2, 8(a2)
>   lw  a7, -12(a5)                   lw  t3, -12(a2)
>   lw  a3, -12(a2)                   lw  t4, 4(a2)
>   add a3, a3, a7                    lw  t5, 0(a2)
>   sw  a3, -12(a2)                   lw  t6, -4(a2)
>   lw  a7, -8(a5)                    lw  s1, -8(a2)
>   lw  a3, -8(a2)                    add s0, t0, a7
>   add a3, a3, a7                    sw  s0, -16(a2)
>   sw  a3, -8(a2)                    lw  s0, -12(a5)
>   lw  a7, -4(a5)                    addi  a4, a4, 8
>   lw  a3, -4(a2)                    add s0, t3, s0
>   add a3, a3, a7                    sw  s0, -12(a2)
>   sw  a3, -4(a2)                    lw  s0, -8(a5)
>   lw  a7, 0(a5)                     add s1, s1, s0
>   lw  a3, 0(a2)                     sw  s1, -8(a2)
>   add a3, a3, a7                    lw  s1, -4(a5)
>   sw  a3, 0(a2)                     add s1, t6, s1
>   lw  a7, 4(a5)                     sw  s1, -4(a2)
>   lw  a3, 4(a2)                     lw  s1, 0(a5)
>   add a3, a3, a7                    add s1, t5, s1
>   sw  a3, 4(a2)                     sw  s1, 0(a2)
>   lw  a7, 8(a5)                     lw  s1, 4(a5)
>   lw  a3, 8(a2)                     add s1, t4, s1
>   add a3, a3, a7                    sw  s1, 4(a2)
>   sw  a3, 8(a2)                     lw  s1, 8(a5)
>   lw  a7, 12(a5)                    add s1, t2, s1
>   lw  a3, 12(a2)                    sw  s1, 8(a2)
>   add a3, a3, a7                    lw  s1, 12(a5)
>   sw  a3, 12(a2)                    addi  a5, a5, 32
>   addi  a4, a4, 8                   add s1, t1, s1
>   addi  a5, a5, 32                  sw  s1, 12(a2)
>   addi  a2, a2, 32                  addi  a2, a2, 32
>   bne a6, a4, .LBB0_4               bne a6, a4, .LBB0_4
>
> @proftaylor <https://github.com/proftaylor> Above is the result I see
> with unrolled loop. To clarify on the scheduling algorithm we override:
>
>    1. Default LLVM scheduler provided two main variants: top-down and
>    bottom-up. We customize default top-down scheduling as opposed to default
>    bottom up because we found bottom-up is bad for non-blocking loads
>    <https://github.com/vb000/llvm-escapade/blob/master/README-notes.md#test-program>.
>    Have you observed the same in GCC?
>    2. It is run before register allocation on IR in SSA form, which means
>    we have a lot of flexibility in reordering instructions.
>    3. This also means the compiler would still reschedule after register
>    allocation based on other heuristics.
>    4. We only override load scheduling --- if there is no load in the top
>    frontier of the control flow graph, we fallback to default top-down
>    scheduling.
>
> Is the counter example you provided still valid?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bespoke-silicon-group/llvm-project/pull/1#issuecomment-639239422>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWTC3LAGZAXEQ3VQXOOL6DRVBSDFANCNFSM4NTFIAAA>
> .
>

vb000 commented:
Thanks a lot for the restrict keyword suggestion!!! I was banging my head to figure out why generated IR couldn't put the loads from the `b` array at the top frontier of the DAG. As you can observe, loads from one array are the top but not loads from the other array. I guess that's the final key to 100% optimal NB load scheduling! I'll take a look at the renaming pass. 

As we have been mentioning in meetings all along, our goal is to achieve desired schedule by just changing the cost model. But in my [experiments](https://github.com/vb000/llvm-escapade/blob/master/README-notes.md#attempt-1-load-latency--2-top-down-scheduling-algorithm-no-register-pressure) (see attempt 1-5) I observed not so promising results by just changing latency. So as a backup, I implemented this custom schedule so that we'd have at least some interesting results to show. In the optimistic scenario, cost model approach works and we would never have to resort to the `-hb32sched` flag I added here. Otherwise, the flag is always there =).

vb000 commented:
`restrict` is in fact the key!!! Output with restrict keyword:
```
.LBB0_4:                                # %for.body
                                        # =>This Inner Loop Header: Depth=1
  lw  a7, -16(a5)
  lw  t0, -16(a2)
  lw  t1, 12(a2)
  lw  t2, 12(a5)
  lw  t3, -12(a5)
  lw  t4, 8(a2)
  lw  t5, -12(a2)
  lw  t6, 8(a5)
  lw  s2, 4(a2)
  lw  s8, -8(a5)
  lw  s0, -8(a2)
  lw  s3, 4(a5)
  lw  s4, 0(a2)
  lw  s5, -4(a5)
  lw  s6, -4(a2)
  lw  s7, 0(a5)
  add s1, t0, a7
  sw  s1, -16(a2)
  add s1, t5, t3
  sw  s1, -12(a2)
  add s1, s0, s8
  sw  s1, -8(a2)
  add s1, s6, s5
  sw  s1, -4(a2)
  add s1, s4, s7
  sw  s1, 0(a2)
  add s1, s2, s3
  sw  s1, 4(a2)
  add s1, t4, t6
  sw  s1, 8(a2)
  add s1, t1, t2
  sw  s1, 12(a2)
  addi  a4, a4, 8
  addi  a5, a5, 32
  addi  a2, a2, 32
  bne a6, a4, .LBB0_4
```

proftaylor commented:
Yay! :-)

On Fri, Jun 5, 2020 at 1:05 AM Bandhav Veluri <notifications@github.com>
wrote:

> restrict is in fact the key!!! Output with restrict keyword:
>
> .LBB0_4:                                # %for.body
>                                         # =>This Inner Loop Header: Depth=1
>   lw  a7, -16(a5)
>   lw  t0, -16(a2)
>   lw  t1, 12(a2)
>   lw  t2, 12(a5)
>   lw  t3, -12(a5)
>   lw  t4, 8(a2)
>   lw  t5, -12(a2)
>   lw  t6, 8(a5)
>   lw  s2, 4(a2)
>   lw  s8, -8(a5)
>   lw  s0, -8(a2)
>   lw  s3, 4(a5)
>   lw  s4, 0(a2)
>   lw  s5, -4(a5)
>   lw  s6, -4(a2)
>   lw  s7, 0(a5)
>   add s1, t0, a7
>   sw  s1, -16(a2)
>   add s1, t5, t3
>   sw  s1, -12(a2)
>   add s1, s0, s8
>   sw  s1, -8(a2)
>   add s1, s6, s5
>   sw  s1, -4(a2)
>   add s1, s4, s7
>   sw  s1, 0(a2)
>   add s1, s2, s3
>   sw  s1, 4(a2)
>   add s1, t4, t6
>   sw  s1, 8(a2)
>   add s1, t1, t2
>   sw  s1, 12(a2)
>   addi  a4, a4, 8
>   addi  a5, a5, 32
>   addi  a2, a2, 32
>   bne a6, a4, .LBB0_4
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bespoke-silicon-group/llvm-project/pull/1#issuecomment-639325263>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWTC3JZUPSGRE3Y7FANK73RVCRK7ANCNFSM4NTFIAAA>
> .
>

vb000 commented:
It may not be the perfect codegen yet... if we take a closer look, load and use are not separated as much as possible for some (t3, t5), but on the other hand, a lot of separation for some (t1, t2). This could be good or bad, only benchmarking will tell.

@sampsyo predicted this might happen in his feedback on this approach. I think, it is possible that this simple customization achieves good results because of random load-use separation. But we still might need to try more predictable policies like "first-load-first-use", "first-load-last-use" or "first-load-random-use". Any thoughts?

proftaylor commented:
Assuming that you are still boosting the priority of the load instruction,
my suggestions are roughly the same as before:

a. Just boosting the priority does not take into account the actual
dependence graph and what operands are higher or lower priority, so the
result will be suboptimal. The default scheduler should be easily able to
handle the case of a load instruction taking 20 cycles; a non-blocking load
is not special from the perspective of a scheduler; it is just an
instruction with a long latency.

b. So I would start should first try just modifying the standard backend
and telling it that the latency of loads is 20 cycles. You can then look at
the code gen to see which scheduler is doing the best job under these
conditions.  My guess is that the schedule will be almost perfect. Then you
have the task of inserting a variable latency for the load based on the C++
attribute.

c. You have three choices; one is to try to modify the scheduler so that it
can compute a conditional cost by looking up the attribute attached to the
memory object. The second choice is to define a new code gen rule for loads
that is the same mapping as the previous rule, but adds an (llvm-internal)
attribute tag to the instruction that the scheduler can use. This is what
worked in GCC for Max. The last option is to create a different instruction
and then pretend that it is different all the way through the backend.
Since Reshabh added support for both 32-bit address loads and 64-bit
address loads which generate entirely different instructions, we know this
is also possible, but it is the most heavy weight approach. In general, to
debug what is going on, you want to see what debugging support is there to
print out scheduling and alias information so you can see what the input
data to the scheduler was, what the algorithm applied is, and compare it
against what you figured out is optimal by hand. This is how you get
insight. The way not to get insight is just to try random things and
benchmark it.


On Fri, Jun 5, 2020 at 11:27 AM Bandhav Veluri <notifications@github.com>
wrote:

> Merged #1 <https://github.com/bespoke-silicon-group/llvm-project/pull/1>
> into hb-dev.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bespoke-silicon-group/llvm-project/pull/1#event-3412630685>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWTC3JXVV4GHVEOOIEJT3DRVE2KLANCNFSM4NTFIAAA>
> .
>

sampsyo commented:
Super cool!

To directly attach the latency of the load, I *think* the right way is to adjust the latency of the `SDep` (dependency edge). As a model, here's a place in the Hexagon backend where they look for specific memory dependencies and do `Dep.setLatency(1)`.
http://llvm.org/doxygen/HexagonSubtarget_8cpp_source.html#l00318

proftaylor commented:
Thanks Adrian!

On Fri, Jun 5, 2020 at 1:26 PM Adrian Sampson <notifications@github.com>
wrote:

> Super cool!
>
> To directly attach the latency of the load, I *think* the right way is to
> adjust the latency of the SDep (dependency edge). As a model, here's a
> place in the Hexagon backend where they look for specific memory
> dependencies and do Dep.setLatency(1).
> http://llvm.org/doxygen/HexagonSubtarget_8cpp_source.html#l00318
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bespoke-silicon-group/llvm-project/pull/1#issuecomment-639786795>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWTC3IM53OR7TEOK5VN4XLRVFIITANCNFSM4NTFIAAA>
> .
>

vb000 commented:
Thanks @sampsyo! Emily and I were trying hard to model dynamic latency confining ourselves to table gen files. While table gen is elegant & recommend way of changing the cost model, it doesn't seems to have features to model dynamic latency. Adjusting latency using `SDep` sounds promising, we'll definitely try that!

Hi @proftaylor, I did try changing the latency to 20 and did see the scheduler logs (llvm has super cool debugging support with `-debug-only=machine-scheduler` flag), which is how I realized the default scheduler's policy might not be very good for our use case. Default scheduling policy _even with 20 cycle delay_ produced sub-par results: please have a look at this: https://github.com/vb000/llvm-escapade/blob/master/README-notes.md#attempt-5-load-latency--20-bottom-up-scheduling-algorithm-no-register-pressure. Default scheduler even with modified load-latency=20 cycles, uses heuristics that are not suitable for NB loads. Those unfavorable heuristics are _bottom-up scheduling_ and _register pressure_. So, **changing latency is only the first step of the problem**. Without right policy, changing latency doesn't produce desired schedules. Having said that, the default scheduler also has various command line options to tweak policy, and we are going to explore those.

In addition to _fully_ exploring above approach, the other point I'd like to make is, critical paths in HB kernels only use a small set of code patterns. But from what I learnt when researching for this, default schedulers (target agnostic schedulers) in compilers are designed keeping super general purpose programs in mind. And given scheduling in an NP-complete problem, I think we may not be able to say one scheduler is optimal and other is not. Especially when target specific knowledge comes into play. From the logs, I see the scheduler highly regards the register pressure when unrolling loops, it never goes beyond using 6 live registers! So, we'd ideally like to say "don't consider register pressure, but only for loads, for other instructions follow you own policy". We can't have these kind of customizations if we are not ready to implement something of our own. A snip from the log, see `MaxPressure`:

```
********** MI Scheduling **********
vec_add:%bb.8 for.body
  From: %29:gpr = LW %65:gpr, -16 :: (load 4 from %ir.scevgep8, !tbaa !3, addrspace 1)
    To: BEQ %1:gpr, %66:gpr, %bb.4
 RegionInstrs: 35
ScheduleDAGMILive::schedule starting
GenericScheduler RegionPolicy:  ShouldTrackPressure=1 OnlyTopDown=0 OnlyBottomUp=1
Max Pressure: GPR=6
Live In: %1 %66 %64 %65
Live Out: %1 %66 %64 %65
Live Thru: GPR=4
  LiveReg: %1
  LiveReg: %66
  LiveReg: %64
  LiveReg: %65
Top Pressure:
GPR=4
Bottom Pressure:
GPR=4
```
Quantitative approach seems to be only reasonable approach to judge different algorithms. I'm not sure if we can say target agnostic heuristics implemented by LLVM  are optimal for HB kernels, even with accurate cost models. So, I think we should not back off from implementing custom algorithms as long as follow a data-driven approach. Curious to hear @proftaylor's and @sampsyo's thoughts...

proftaylor commented:
But in that link you sent, did you use the restrict keyword?

On Fri, Jun 5, 2020 at 2:55 PM Bandhav Veluri <notifications@github.com>
wrote:

> Thanks @sampsyo <https://github.com/sampsyo>! Emily and I were trying
> hard to model dynamic latency confining ourselves to table gen files. While
> table gen is elegant & recommend way of changing the cost model, it doesn't
> seems to have features to model dynamic latency. Adjusting latency using
> SDep sounds promising, we'll definitely try that!
>
> Hi @proftaylor <https://github.com/proftaylor>, I did try changing the
> latency to 20 and did see the scheduler logs (llvm has super cool debugging
> support with -debug-only=machine-scheduler flag), which is how I realized
> the default scheduler's policy might not be very good for our use case.
> Default scheduling policy *even with 20 cycle delay* produced sub-par
> results: please have a look at this:
> https://github.com/vb000/llvm-escapade/blob/master/README-notes.md#attempt-5-load-latency--20-bottom-up-scheduling-algorithm-no-register-pressure.
> Default scheduler even with modified load-latency=20 cycles, uses
> heuristics that are not suitable for NB loads. Those unfavorable heuristics
> are *bottom-up scheduling* and *register pressure*. So, *changing latency
> is only the first step of the problem*. Without right policy, changing
> latency doesn't produce desired schedules. Having said that, the default
> scheduler also has various command line options to tweak policy, and we are
> going to explore those.
>
> In addition to *fully* exploring above approach, the other point I'd like
> to make is, critical paths in HB kernels only use a small set of code
> patterns. But from what I learnt when researching for this, default
> schedulers (target agnostic schedulers) in compilers are designed keeping
> super general purpose programs in mind. And given scheduling in an
> NP-complete problem, I think we may not be able to say one scheduler is
> optimal and other is not. Especially when target specific knowledge comes
> into play. From the logs, I see the scheduler highly regards the register
> pressure when unrolling loops, it never goes beyond using 6 live registers!
> So, we'd ideally like to say "don't consider register pressure, but only
> for loads, for other instructions follow you own policy". We can't have
> these kind of customizations if we are not ready to implement something of
> our own. A snip from the log, see MaxPressure:
>
> ********** MI Scheduling **********
> vec_add:%bb.8 for.body
>   From: %29:gpr = LW %65:gpr, -16 :: (load 4 from %ir.scevgep8, !tbaa !3, addrspace 1)
>     To: BEQ %1:gpr, %66:gpr, %bb.4
>  RegionInstrs: 35
> ScheduleDAGMILive::schedule starting
> GenericScheduler RegionPolicy:  ShouldTrackPressure=1 OnlyTopDown=0 OnlyBottomUp=1
> Max Pressure: GPR=6
> Live In: %1 %66 %64 %65
> Live Out: %1 %66 %64 %65
> Live Thru: GPR=4
>   LiveReg: %1
>   LiveReg: %66
>   LiveReg: %64
>   LiveReg: %65
> Top Pressure:
> GPR=4
> Bottom Pressure:
> GPR=4
>
> Quantitative approach seems to be only reasonable approach to judge
> different algorithms. I'm not sure if we can say target agnostic heuristics
> implemented by LLVM are optimal for HB kernels, even with accurate cost
> models. So, I think we should not back off from implementing custom
> algorithms as long as follow a data-driven approach. Curious to hear
> @proftaylor <https://github.com/proftaylor>'s and @sampsyo
> <https://github.com/sampsyo>'s thoughts...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bespoke-silicon-group/llvm-project/pull/1#issuecomment-639843991>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWTC3O3SHATFNXVVKZYZ5DRVFSULANCNFSM4NTFIAAA>
> .
>

vb000 commented:
Ah, that doesn't have it and when I run it with restrict, it seems to generate good schedules and disregard register pressure:
```
.LBB0_4:

  lw  a7, -8(a5)
  lw  t0, -8(a2)
  lw  t1, 4(a2)
  lw  t2, 4(a5)
  lw  t3, 0(a2)
  lw  t4, 0(a5)
  lw  t5, -4(a2)
  lw  t6, -4(a5)
  addi  a4, a4, 4
  addi  a5, a5, 16
  add a3, t0, a7
  sw  a3, -8(a2)
  add a3, t1, t2
  sw  a3, 4(a2)
  add a3, t3, t4
  sw  a3, 0(a2)
  add a3, t5, t6
  sw  a3, -4(a2)
  addi  a2, a2, 16
  bne a6, a4, .LBB0_4
```

Guess the discussion above is moot point now. Changing the load-latency seems to work perfectly!

proftaylor commented:
Earlier, as part of the “hypothesis” step I was talking about, I wrote down
the schedule that I would use for hand tuned code, and it looked identical
to what you just wrote. Interleaving the adds and stores maximizes the
load-to-use distance. So this is a very good schedule indeed.

On Fri, Jun 5, 2020 at 5:55 PM Bandhav Veluri <notifications@github.com>
wrote:

> Ah, that doesn't have it and when I run it with restrict, it seems to
> generate good schedules and disregard register pressure:
>
> .LBB0_4:
>
>   lw  a7, -8(a5)
>   lw  t0, -8(a2)
>   lw  t1, 4(a2)
>   lw  t2, 4(a5)
>   lw  t3, 0(a2)
>   lw  t4, 0(a5)
>   lw  t5, -4(a2)
>   lw  t6, -4(a5)
>   addi  a4, a4, 4
>   addi  a5, a5, 16
>   add a3, t0, a7
>   sw  a3, -8(a2)
>   add a3, t1, t2
>   sw  a3, 4(a2)
>   add a3, t3, t4
>   sw  a3, 0(a2)
>   add a3, t5, t6
>   sw  a3, -4(a2)
>   addi  a2, a2, 16
>   bne a6, a4, .LBB0_4
>
> Guess the discussion above is moot point now. Changing the load-latency
> seems to work perfectly!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bespoke-silicon-group/llvm-project/pull/1#issuecomment-639928712>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWTC3PVGADWMGAMYXHHUWTRVGHWXANCNFSM4NTFIAAA>
> .
>

vb000 commented:
That's right, scheduler is simulating the processor based on the cost model, and tries to minimize the heuristic "Expected cycles". In this particular case, when scheduler tries `add` followed by `store`, it doesn't see a change in the running count of simulated cycles, because scheduler accounted 20 cycles for the first load!

And not to mention, resolving pointer aliasing is what makes all of this possible. Without that, basically there's not much ILP possible in the test program we are looking at. A nice learning in how an instruction scheduler design should be thought of!
